crates/sui-types/src/temporary_store.rs

```diff
@@ -689,6 +689,150 @@ impl<S> TemporaryStore<S> {
+                            // Note: this assumes that the only immutable objects an epoch change tx can update are system packages,
+                            // but in principle we could allow others.
+                            assert!(
+                                is_system_package(*id),
+                                "Only system packages can be upgraded"
+                            );
+                        }
+                    }
+                }
+                WriteKind::Create | WriteKind::Unwrap => {
+                    // created and unwrapped objects were not inputs to the tx
+                    // or dynamic fields accessed at runtime, no ownership checks needed
+                }
+            }
+        }
+
+        for (id, (_version, kind)) in &self.deleted {
+            if authenticated_objs.contains(id) || gas_objs.contains(id) {
+                continue;
+            }
+            match kind {
+                DeleteKind::Normal | DeleteKind::Wrap => {
+                    // get owner at beginning of tx
+                    let old_obj = self.store.get_object(id)?.unwrap();
+                    match &old_obj.owner {
+                        Owner::ObjectOwner(_) => {
+                            objs_to_authenticate.push(*id);
+                        }
+                        Owner::AddressOwner(_) | Owner::Shared { .. } => {
+                            unreachable!("Should already be in authenticated_objs")
+                        }
+                        Owner::Immutable => unreachable!("Immutable objects cannot be deleted"),
+                    }
+                }
+                DeleteKind::UnwrapThenDelete => {
+                    // unwrapped-then-deleted object was not an input to the tx,
+                    // no ownership checks needed
+                }
+            }
+        }
+        Ok((objs_to_authenticate, authenticated_objs))
+    }
+
+    // check that every object read is owned directly or indirectly by sender, sponsor, or a shared object input
+    pub fn check_ownership_invariants(
+        &self,
+        sender: &SuiAddress,
+        gas: &[ObjectRef],
+        is_epoch_change: bool,
+    ) -> SuiResult<()> {
+        let (mut objects_to_authenticate, mut authenticated_objects) =
+            self.get_objects_to_authenticate(sender, gas, is_epoch_change)?;
+
+        // Map from an ObjectID to the ObjectID that covers it.
+        let mut covered = BTreeMap::new();
+        // iterate until we we have authenticated all objects, or reached a fixed point with remaining objects to authenticate
+        while let Some(to_authenticate) = objects_to_authenticate.pop() {
+            let old_obj = self.store.get_object(&to_authenticate)?.expect(
+                "Object authentication failure: could not find parent object {id:?},\n\
+            which must be authenticated to cover one or more objects in {covered:#?}.",
+            );
```

Comment:
The formatting string won't work here (noticed from the error output you mentioned), how about the following:

```suggestion
            let Some(old_obj) = self.store.get_object(&to_authenticate)? else {
                panic!(
                    "Object authentication failure: could not load object {to_authenticate:?},\n\
                     which may cover one or more objects in {covered:#?}.",
                );
            }
```

Also tweaked the wording (objects in `to_authenticate` may not necessarily be covering anything).